general/src/pii/processor.rs

```diff
@@ -214,66 +212,141 @@ impl<'a, 'b> Iterator for RuleIterator<'a, 'b> {
     }
 }
 
+macro_rules! value_process_method {
+    ($name: ident, $ty:ident $(::$path:ident)*) => {
+        value_process_method!($name, $ty $(::$path)* <>);
+    };
+
+    ($name: ident, $ty:ident $(::$path:ident)* < $($param:ident),* > $(, $param_req_key:ident : $param_req_trait:path)*) => {
+        #[inline]
+        fn $name<$($param),*>(
+            &mut self,
+            value: &mut $ty $(::$path)* <$($param),*>,
+            meta: &mut Meta,
+            state: &ProcessingState<'_>,
+        ) -> ValueAction
+        where
+            $($param: ProcessValue),*
+            $(, $param_req_key : $param_req_trait)*
+        {
+            let mut rules = self.iter_rules(state).peekable();
+            let rv = if rules.peek().is_some() {
+                value_process(value, meta, rules)
+            } else {
+                ValueAction::Keep
+            };
+            match rv {
+                ValueAction::Keep => ValueAction::Keep,
+                other => {
+                    value.process_child_values(self, state);
+                    other
+                }
+            }
+        }
+    };
+}
+
+fn value_process<'a, T: ProcessValue, I: Iterator<Item = RuleRef<'a>>>(
+    value: &mut T,
+    meta: &mut Meta,
+    rules: I,
+) -> ValueAction {
+    for rule in rules {
+        match apply_rule_to_value(value, meta, rule, None) {
+            ValueAction::Keep => continue,
+            other => return other,
+        }
+    }
+    ValueAction::Keep
+}
+
 impl<'a> Processor for PiiProcessor<'a> {
+    value_process_method!(process_i64, i64);
+    value_process_method!(process_u64, u64);
+    value_process_method!(process_f64, f64);
+    value_process_method!(process_bool, bool);
+    value_process_method!(process_timestamp, Timestamp);
+    value_process_method!(process_value, crate::types::Value);
+    value_process_method!(process_array, crate::types::Array<T>);
+
     fn process_string(
         &mut self,
-        string: &mut String,
+        value: &mut String,
         meta: &mut Meta,
         state: &ProcessingState<'_>,
     ) -> ValueAction {
-        let mut rules = self.iter_rules(PiiKind::Text, &state).peekable();
+        let mut rules = self.iter_rules(state).peekable();
         if rules.peek().is_some() {
-            process_chunked_value(string, meta, |chunks| {
-                rules.fold(chunks, apply_rule_to_chunks)
+            let rules: SmallVec<[RuleRef; 16]> = rules.collect();
```

Comment:
_nit_: Since collecting an empty vector is insanely cheap and there is not much optimization with the `SmallVec` here, we might be able to simplify this part.